File: scripts/install-configure_RHEL65_yum_repo.rb
# Cookbook Name:: bws_9_dot_9_telus
# Recipe:: install-configure-RHEL65_yum_repo_server.rb
#    
#
# Copyright 2016, Amdocs
#
# All rights reserved - Do Not Redistribute
#

########################################################################
## make the /chef/rhel65_yum_repo directory 
########################################################################
directory "#{node.bws.config.rhel65.iso_dest_dir}" do
  owner 'root'
  group 'root'
  mode '0755'
end


#########################################################################
## get the RHEL65 image to the right place
########################################################################
remote_file "Copy RHEL 6.5 ISO" do
  path "#{node.bws.config.rhel65.iso_dest_dir}#{node.bws.config.rhel65.iso_name}"
  source "#{node.bws.config.rhel65.iso_src_dir}#{node.bws.config.rhel65.iso_name}"
  owner 'root'
  group 'root'
  mode 0755
  action :create_if_missing
end


#########################################################################
## make the /var/www/html/cdrom/iso directory tree  (assume /var is already there)
########################################################################
directory '/var/www/' do
  owner 'root'
  group 'root'
  mode '0755'
end

directory '/var/www/html/' do
  owner 'root'
  group 'root'
  mode '0755'
end

directory '/var/www/html/cdrom/' do
  owner 'root'
  group 'root'
  mode '0755'
end

directory '/var/www/html/cdrom/iso/' do
  not_if {::File.exist? '/var/www/html/cdrom/iso/'}
  owner 'root'
  group 'root'
  mode '0755'
  ignore_failure true
end

#########################################################################
## mount it!
########################################################################
mount '/var/www/html/cdrom/iso/' do
  device "#{node.bws.config.rhel65.iso_dest_dir}#{node.bws.config.rhel65.iso_name}"
  fstype 'iso9660'
  pass 0
  options ' loop'
  action [:mount, :enable]
end


########################################################################
## import the GPG key locally
#######################################################################
execute 'installing RHEL 6.5 GPG key' do
  command "rpm --import /var/www/html/cdrom/iso/RPM-GPG-KEY-redhat-release"
  creates ''
  action :run
end


########################################################################
## install the deltarpm package
########################################################################
execute 'installing deltarpm' do
  command "rpm -i /var/www/html/cdrom/iso/Packages/deltarpm*.x86_64.rpm"
  creates '/usr/bin/applydeltarpm'
  action :run
end

########################################################################
## install the python-deltarpm package
########################################################################
execute 'installing python-deltarpm' do
  command "rpm -i /var/www/html/cdrom/iso/Packages/python-deltarpm*.x86_64.rpm"
  creates '/usr/lib64/python2.6/site-packages/deltarpm.py'
  action :run
end

########################################################################
## install createrepo from the ISO image
#######################################################################
execute 'installing createrepo' do 
  command "rpm -i /var/www/html/cdrom/iso/Packages/createrepo*.noarch.rpm"
  creates '/usr/bin/createrepo'
  action :run
end

########################################################################
## create the repomd from the image
#######################################################################
execute 'creating the repomd data for the ISO' do
  command 'createrepo /var/www/html/cdrom/'
  creates '/var/www/html/cdrom/iso/repodata'
end


template "/etc/yum.repos.d/rhel65.repo" do
    source "rhel65.repo.erb"
    variables(
        :baseurl => node.bws.config.rhel65.yum_repo_url
    ) 
    owner "root"
    group "root"
    mode "0644"
    action :create
end

File: scripts/Python/install-configure_RHEL65_yum_repo.py
import paramiko
from cloudshell.helpers.scripts import cloudshell_scripts_helpers as helper
import re

def _ssh_write(ssh, channel, command):

    channel.send(command)



def _ssh_read(ssh, channel, prompt_regex):

    rv = ''

    while True:
        # self.channel.settimeout(30)

        r = channel.recv(2048)

        if r:
            rv += r
        if rv:
            t = rv
            t = re.sub(r'(\x9b|\x1b)[[?;0-9]*[a-zA-Z]', '', t)
            t = re.sub(r'(\x9b|\x1b)[>=]', '', t)
            t = re.sub('.\b', '', t)  # not r''
        else:
            t = ''
        if not r or len(re.findall(prompt_regex, t)) > 0:
            rv = t
            if rv:
                rv = rv.replace('\r', '\n')

            return rv


def _ssh_command(ssh, channel, command, prompt_regex):

    _ssh_write( ssh, channel, command + '\n')
    rv = _ssh_read( ssh, channel, prompt_regex)
    if '\n%' in rv.replace('\r', '\n'):
        es = 'CLI error message: ' + rv

        raise Exception(es)
    return rv

resource_context = helper.get_resource_context_details()
session = paramiko.SSHClient()
session.set_missing_host_key_policy(paramiko.AutoAddPolicy())
session.connect(resource_context.address, 22, 'root','amdocs')

channel = session.invoke_shell()
prompt = '.*]#'
_ssh_command(session, channel, 'cd /stage', prompt)
_ssh_command(session, channel, 'mkdir -p /var/www/html/cdrom/iso', prompt)
_ssh_command(session, channel, 'mount -o loop /stage/rhel-server-5.8-x86_64-dvd.iso /var/www/html/cdrom/iso', prompt)
_ssh_command(session, channel, 'rpm --import /var/www/html/cdrom/iso/RPM-GPG-KEY-redhat-release',prompt)
_ssh_command(session, channel, 'yum install deltarpm-3.5+0.5.20090913git.e16.x86_64.rpm python-deltarpm-3.5-0.5.20090913git.e16.x86)64.rpm','.*y/N.*')
_ssh_command(session, channel, 'y', prompt)
_ssh_command(session, channel, 'rpm -i /var/www/html/cdrom/iso/Packages/createrepo*.noarch.rpm', prompt)
_ssh_command(session, channel, 'cd /var/www/html/cdrom', prompt)
_ssh_command(session, channel, 'createrepo .', prompt)
_ssh_command(session, channel, 'rm -rf /etc/yum.repos.d/rhel-source.repo', prompt)
_ssh_command(session, channel, 'echo [RHEL-Repository] > /etc/yum.repos.d/file.repo', prompt)
_ssh_command(session, channel, 'echo name=RHEL repository >>  /etc/yum.repos.d/file.repo', prompt)
_ssh_command(session, channel, 'echo baseurl=file:///var/www/html/cdrom >> /etc/yum.repos.d/file.repo', prompt)
_ssh_command(session, channel, 'echo enabled=1 >> /etc/yum.repos.d/file.repo', prompt)
_ssh_command(session, channel, 'echo gpgcheck=0 >> /etc/yum.repos.d/file.repo', prompt)

_ssh_command(session, channel, 'yum clean all', prompt)


